File: projects/06/assembler.py
import parser
import sys
import os
import re

class Assembler:
    def __init__(self, argv=sys.argv[1]):
        # split into filename and extenstion 
        self.pre, ext = os.path.splitext(argv)
        f = open(self.pre + ext, 'r') 

        # create input stream
        self.input = f.read()

        # clean code: remove blank lines and split into list 
        comment_free = re.sub('(/\*([^*]|[\r\n]|(\*+([^*/]|[\r\n])))*\*+/)|(//.*)|[^\S\r\n]', '', self.input) # removes comments
        lines = comment_free.splitlines() # splits into list
        self.input = [x for x in lines if x] # removes empty items / blank lines from list

        # close file
        f.close()

        # instantiate objects
        self.parser = parser.Parser(self.input)
        self.code = parser.Code()
        self.symbol_table = parser.SymbolTable()
        self.addr = 16

        #  First-pass through input stream
        # We loop through input stream and add L_COMMANDS and binary index to Symbol Table
        input_dupe = [] 

        # First pass through input file. Remove L_COMMANDS and add Variables to symbol_table
        while self.parser.hasMoreCommands():
            self.parser.advance()
            command = self.parser.getCommand()

            if self.parser.commandType(command) == 'L_COMMAND':
                address = "{0:016b}".format(len(input_dupe)) # convert index position to binary location
                self.symbol_table.addEntry(command[1:-1], address) # note, first and last char '(' and ')' sliced from command

            else:
                input_dupe.append(command)

        self.parser.input = input_dupe.copy() # replace original list with duplicate


    def parse(self, line):
        """
        Parses an ASM command and outputs a corrosponding HACK file
        Input: ASM command (str)
        Returns: 16 digit binary (Hack) command (str)
        Note: code is not error checking. Assumes all ASM commands are correct 
        """
        
        if self.symbol_table.contains(line[1:]):
            return(self.symbol_table.getAddress(line[1:]))

        elif self.parser.commandType(line) == 'A_COMMAND' and line[1].isalpha(): # Command is variable reference
            loc = "{0:016b}".format(self.addr)
            self.symbol_table.addEntry(line[1:], loc)
            self.addr += 1
            return loc

        elif self.parser.commandType(line) == 'A_COMMAND':
            if self.symbol_table.contains(line[1:]):
                return self.symbol_table.getAddress(line[1:])
            else:
                return "{0:016b}".format(int(line[1:]))
                
        elif self.parser.commandType(line) == 'C_COMMAND':
            ins = '111'

            self.parser.setCommand(line)
            mnem = self.parser.comp()
            comp = self.code.comp(mnem)

            mnem = self.parser.dest()
            dest = self.code.dest(mnem)

            mnem = self.parser.jump()
            try:
                jump = self.code.jump(mnem)
            except:
                jump = '000'

            return ins + comp + dest + jump

    # Write input file to output
    def parse_file(self):
        f = open(self.pre + '1.hack', 'w') # create output file

        while self.parser.hasMoreCommands():
            self.parser.advance() # pops first item to parsers current command
            hack_line = self.parse(self.parser.getCommand()) # parse it to hack command
            f.write(hack_line + '\n') # write to file
        f.close() # Close output file
    
assembler = Assembler(sys.argv[1])
assembler.parse_file()

File: projects/06/Assembler/src/Assembler.java
import java.util.*;
import java.io.*;
import java.nio.file.Path;

public class Assembler {
	// HashMap<String, Integer> symbolTable = new HashMap<String, Integer>();
	SymbolTable symbolTable = new SymbolTable();
	Parser parser = new Parser();	
	Code code = new Code();
	ArrayList<String> output = new ArrayList<String>();
	
	public static void main(String[] args) {
		
		Assembler assembler = new Assembler();
		
		String asm = new String();
		if(args.length >= 0) {
			asm = args[0];

			assembler.firstPass(asm);			
			assembler.secondPass();
			// assembler.output()

		} else {
			System.out.println("Missing argument: ASM file");
		}

		assembler.firstPass(asm);
		assembler.secondPass();
		
	}

	public String intToBin(Integer num) {
		String b = String.format("%16s", Integer.toBinaryString(num)).replace(' ', '0');
		return(b);

	}

	public String stringToBin(String num) {
		Integer i = Integer.parseInt(num);
		String b = String.format("%16s", Integer.toBinaryString(i)).replace(' ', '0');
		return(b);

	}

	public void firstPass(String asm) {
		parser.initialize(asm);
		symbolTable.initialize();
		
		int lineNumber = 0;
		ArrayList<String> inputWithoutL = new ArrayList<String>(); // where we hold all commands bar L_COMMANDs

		while(parser.hasMoreCommands()) {
			parser.advance();
			
			if(parser.commandType() == "L_COMMAND") {
				symbolTable.addEntry(parser.symbol(), lineNumber);
			} else {
				inputWithoutL.add(parser.currentCommand);
				lineNumber += 1;
			}
		}
		// parser.input.clear();
		parser.input.addAll(inputWithoutL);
	}
	
	public void secondPass() {

		while( parser.hasMoreCommands() ) {
			parser.advance();
			Integer symbolTableVal = 16;
			
			if( parser.commandType() == "A_COMMAND" ) {
				String currentSymbol = parser.symbol();

				if ( !parser.isNumeric() ) {
					if ( symbolTable.contains(currentSymbol) ) {
						Integer key = symbolTable.getAddress(currentSymbol);
						String binary = intToBin(key);
						output.add(binary);
					} 
					else {
						symbolTable.addEntry(parser.symbol(), symbolTableVal);
						symbolTableVal += 1;
					}
				} 
				
				else {

					int i = Integer.parseInt(currentSymbol);
					String b = String.format("%16s", Integer.toBinaryString(i)).replace(' ', '0');
					
					output.add(b);	
				}

			}

			else if (parser.commandType() == "C_COMMAND") {
				
				String d = code.dest(parser.dest());
				String c = code.comp(parser.comp());
				String j = code.jump(parser.jump());

				output.add("111" + c + d + j);
			}

			else if (parser.commandType() == "L_COMMAND") {

			}

			else {
				String currentSymbol = parser.symbol();
				output.add(currentSymbol);
			}
		}
	}

	// Output
	public void output(String fileName) {
		String[] tokens = fileName.split("\\.(?=[^\\.]+$)");
		String path = tokens[0];

		try {
			File file = new File(path + "1.hack");
			file.createNewFile();
			System.out.println("File" + file); 
		} catch(Exception ex) {
			ex.printStackTrace();
		}
		
		// implement FileWriter https://www.tutorialspoint.com/java-program-to-write-int-array-to-a-file
	}
		
	

}
